Sharepoint integration

masaball opened the pull request:
This work adds basic Sharepoint integration to browse_everything. It uses the "[access on behalf of a user](https://learn.microsoft.com/en-us/graph/auth-v2-user?tabs=http)" authorization flow to allow a user to authenticate to the Microsoft Graph API and upload files from their personal and joined Team drives.

masaball committed:
Bring in sharepoint code

Refer to https://github.com/samvera/browse-everything/pull/418 for
what is being built off of.

masaball committed:
Implement authorization_code flow

masaball committed:
Update SharePoint.md

masaball commented on SharePoint.md:
```diff
@@ -0,0 +1,28 @@
+
+Prerequisite:
+  * App must be registered in the Entra Admin center to receive client_id, client_secret, and tenant_id.
+  * If using .default endpoint as your scope, you must register API permissions for your application. Minimum permissions:
```
May also require `Sites.Selected` when implementing sites. Will need to test .

masaball replied:
`Sites.Selected` is not necessary.

masaball commented on lib/browse_everything/driver/sharepoint.rb:
```diff
@@ -0,0 +1,209 @@
+        base.each do |k,v|
+          query += ["#{k}=#{v}"]
+        end
+        query += ["response_type=code"]
```
Rather than initializing `query` as an empty array, this could be the initialization value instead of adding it later.

If/when we want to commit this back upstream we will want to add back in the code paths for the [access without a user](https://learn.microsoft.com/en-us/graph/auth-v2-service?tabs=http) auth flow so leaving the initialization as is, with the expectation of having multiple login options may be preferable. (Access without a user requites the client secret to be sent with the auth request and does not send a `response_type` param).

masaball committed:
Return all user accessible sharepoint sites

masaball committed:
Add an option to filter what sites get returned

masaball committed:
Use select param to limit returned metadata

We do not use every metadata field returned from the all sites and all
drives queries. Using the select parameter to limit what gets returned
to the handful of fields we need should increase the speed of the
requests, resulting in faster loading.

masaball committed:
Update documentation

masaball committed:
Merge pull request #3 from avalonmediasystem/sharepoint_sites

Add support for sites to sharepoint integration

cjcolvar commented on lib/browse_everything/driver/sharepoint.rb:
```diff
@@ -0,0 +1,232 @@
+
+      def auth_query_string
+        query = []
+        # keep_if deletes from and returns self, so dup config to not overwrite original
+        base = config.dup.keep_if { |k,v| ['client_id', 'scope', 'redirect_uri'].include?(k) }
```
I think `slice` does the same thing as this line.
```suggestion
        base = config.slice('client_id', 'scope', 'redirect_uri')
```

cjcolvar commented on lib/browse_everything/driver/sharepoint.rb:
```diff
@@ -0,0 +1,232 @@
+
+      def expiration_time
+        return unless @token
+        @token.fetch('expires_at', nil).to_i
```
I think you'll need to add special handling for nil because `nil.to_i == 0` which would cause the `expiration_time.nil?` check in `.token_expired?` fail.  Maybe something like this?
```suggestion
        expires_at = @token.fetch('expires_at', nil)
        expires_at.nil? ? nil : expires_at.to_i
```

cjcolvar reviewed (commented):
Overall this looks good as far as I can tell.  Not knowing the browse-everything code very well, I wonder if all of the methods are getting used or if there are some that could be removed.  Also I think it would be good to add some tests for the driver if that wouldn't be too hard and seem like they're actually testing something.

masaball committed:
Add tests for sharepoint driver

This commit also includes some fixes to take care of the bulk of errors
from rubocop.

cjcolvar commented on spec/lib/browse_everything/driver/sharepoint_spec.rb:
```diff
@@ -0,0 +1,260 @@
+
+      it 'exposes the authorization endpoint URI' do
+        expect(uri).to be_a Addressable::URI
+        expect(uri.to_s).to eq 'https://login.microsoftonline.com/TENANTID/oauth2/v2.0/authorize?client_id=CLIENTID&scope=offline_access https://graph.microsoft.com/.default&redirect_uri=http://example.com/browse/connect&response_type=code'
```
This looks odd.  Are there supposed to be two urls?  From what I can tell Addressable::URI includes the second url in the query string of the first.

masaball replied:
It is a single URL. Scopes are space separated like `&scope=offline_access%20user.read%20mail.read`. So this URI will be `https://login.microsoftonline.com/TENANTID/oauth2/v2.0/authorize?client_id=CLIENTID&scope=offline_access%20https://graph.microsoft.com/.default&redirect_uri=http://example.com/browse/connect&response_type=code` when it gets encoded. 

The scope pointing to that graph URL is used so that we can assign the required permissions through the microsoft entra admin center rather than having to hardcode the permissions or have them as yet another configuration option in Avalon.

cjcolvar replied:
:+1:  What do you think about changing this line slightly so that's a little more explicit?
```suggestion
        expect(uri.normalize.to_s).to eq 'https://login.microsoftonline.com/TENANTID/oauth2/v2.0/authorize?client_id=CLIENTID&scope=offline_access%20https://graph.microsoft.com/.default&redirect_uri=http://example.com/browse/connect&response_type=code'
```

masaball committed:
Enable permissions re-consent flow

masaball committed:
Use joinedTeams as base folder instead of sites

The sites endpoint returns every single sharepoint site that a user has
access to. This list can include many more options than a user would
expect, including stuff that the user should technically not have access
to (such as if a site has its permissions misconfigured). While we can
use a $search param to narrow the results, we can only get so granular
and so even the filtered results may provide an unwieldy amount of
folders.

Because IU is interested in uploading files that are stored on
sharepoint drives that originate from Microsoft Teams, we can use the
`joinedTeams` endpoint to only return drives that a user is explicitly a
member of. This creates a much more sensible and manageable list for
users to browse through and one that should have no surprises for the
user.

This commit also includes some Rubcocop rule management changes to get
everything returning green.

masaball committed:
Merge pull request #4 from avalonmediasystem/sharepoint_teams

Change base folder path to user's joinedTeams instead of Sites

cjcolvar reviewed (approved):
This looks great!  

Now that the focus has shifted from drives (filtered) to teams I think it would be good to highlight that in `SharePoint.md` to make it explicit what specific part of SharePoint is being exposed in this adapter.

masaball committed:
Update SharePoint.md

cjcolvar commented:
Looks good!
